test/Atomics/AtomicsFolding.swift.gyb

```diff
@@ -0,0 +1,562 @@
+    desired: desired,
+    ordering: .acquiring)
+}
+// SECTIONING-LABEL: } // end sil function '$s14AtomicsFolding
+
+// SECTIONING-LABEL: // compareExchange_${label}_releasing(_:_:_:)
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public func compareExchange_${label}_releasing(
+  _ value: UnsafeAtomic<${type}>,
+  _ expected: ${type},
+  _ desired: ${type}
+) -> (exchanged: Bool, original: ${type})  {
+  // EXCLUSIONS-NOT: builtin "cmpxchg_monotonic_{{[a-z0-9_]*}}_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_acquire_{{[a-z0-9_]*}}_${builtin}"
+  // CHECK: builtin "cmpxchg_release_monotonic_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_acqrel_{{[a-z0-9_]*}}_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_seqcst_{{[a-z0-9_]*}}_${builtin}"
+  return value.compareExchange(
+    expected: expected,
+    desired: desired,
+    ordering: .releasing)
+}
+// SECTIONING-LABEL: } // end sil function '$s14AtomicsFolding
+
+// SECTIONING-LABEL: // compareExchange_${label}_acquiringAndReleasing(_:_:_:)
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public func compareExchange_${label}_acquiringAndReleasing(
+  _ value: UnsafeAtomic<${type}>,
+  _ expected: ${type},
+  _ desired: ${type}
+) -> (exchanged: Bool, original: ${type})  {
+  // EXCLUSIONS-NOT: builtin "cmpxchg_monotonic_{{[a-z0-9_]*}}_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_acquire_{{[a-z0-9_]*}}_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_release_{{[a-z0-9_]*}}_${builtin}"
+  // CHECK: builtin "cmpxchg_acqrel_acquire_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_seqcst_{{[a-z0-9_]*}}_${builtin}"
+  return value.compareExchange(
+    expected: expected,
+    desired: desired,
+    ordering: .acquiringAndReleasing)
+}
+// SECTIONING-LABEL: } // end sil function '$s14AtomicsFolding
+
+// SECTIONING-LABEL: // compareExchange_${label}_sequential(_:_:_:)
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public func compareExchange_${label}_sequential(
+  _ value: UnsafeAtomic<${type}>,
+  _ expected: ${type},
+  _ desired: ${type}
+) -> (exchanged: Bool, original: ${type})  {
+  // EXCLUSIONS-NOT: builtin "cmpxchg_monotonic_{{[a-z0-9_]*}}_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_acquire_{{[a-z0-9_]*}}_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_release_{{[a-z0-9_]*}}_${builtin}"
+  // EXCLUSIONS-NOT: builtin "cmpxchg_acqrel_{{[a-z0-9_]*}}_${builtin}"
+  // CHECK: builtin "cmpxchg_seqcst_seqcst_${builtin}"
+  return value.compareExchange(
+    expected: expected,
+    desired: desired,
+    ordering: .sequentiallyConsistent)
+}
+// SECTIONING-LABEL: } // end sil function '$s14AtomicsFolding
+
+% end
+
+% for (operation, modifier) in [("compareExchange", ""), ("weakCompareExchange", "_weak")]:
+// -----------------------------------------------------------------------------
+// FIXME: There are 15 valid combinations of success/failure orderings;
+// here we're only testing 6 of them.
```

Comment:
Is this intended to be fixed before merging? I see there's a

```
FIXME: Swift should provide intrinsics for arbitrary ordering pairs
```

at the beginning -- I'm not entirely sure if the two are related. If they are, please add a note or something indicating the relation between the two. For example, you could have something like `FIXME: [limited-ordering-pairs]` in both places so one can easily connect the two.

Reply:
> Is this intended to be fixed before merging?

Maybe -- the obvious way to add the missing test cases would be to stamp out all the combinations with gyb, but this test is already unreadably meta. It's not clear that adding full coverage is necessary here; I can't imagine the compiler would decide to arbitrarily refuse to optimize just one of the argument combinations. But what do I know... ¯\_(ツ)_/¯

> I'm not entirely sure if the two are related.

They aren't! We can stop "maxing" success ordering to the failure ordering when we switch to an llvm version that supports independent success/failure orderings. (I believe this might be on the way, since C++ looks set to lift the corresponding constraint soon.)

Reply:
> Maybe -- the obvious way to add the missing test cases would be to stamp out all the combinations with gyb, but this test is already unreadably meta. It's not clear that adding full coverage is necessary here; I can't imagine the compiler would decide to arbitrarily refuse to optimize just one of the argument combinations. But what do I know... ¯\\_(ツ)_/¯

Ok... there's two arguments here. (1) It is better to be safe than sorry (so add the missing combinations) and (2) The value add of those combinations is negligible (so not add the missing combinations). Personally I would lean towards (1), but I can understand if you think (2) is more important here.

In any case, leaving the FIXME kinda' makes it seems like there is something to be fixed, but that doesn't seem to be the intention. So my thought here is:

1. We get rid of the FIXME and adding the missing combinations.
2. OR We add a reason for why the combinations are missing, and remove the FIXME, just have it be a normal comment (indicating that it is deliberate decision and not for lack of time).

Reply:
FIXMEs do not come with an expectation that we need (or even want) them to be fixed -- they merely note an aspect that we could perhaps revisit, if we had no time/resource/ABI constraints.

The test serves it intended purpose as it is! Adding test coverage for all ordering combinations is non-essential work, and we can easily do it (or not) at any point. The FIXME does its job by calling out the missing combos.

Short term, it would probably be more important to work on adding different tests, especially tests that actually exercise atomicity and memory orderings.